### PR TITLE
Retire the rewrite to BRGEMM pass

### PR DESCRIPTION
**Pull request by chelini**
Fusion now is general enough and the tiling strategy used in tile and
fuse already exposes BRGEMMs. Thus, `rewrite-to-brgemm` becomes a simple
trivial rewrite. Adjust fusion to tile if the working group consists of
a single operation, we need to do this as we benchmarks and map 
single operations (for single op we tile only). 
Previously, fusion was assuming a working group
of at least two operations.

Fixes: https://github.com/plaidml/tpp-mlir/issues/392

### CODE REVIEW

File: lib/TPP/LinalgDeGeneralize.cpp [hunk 1]
```diff
@@ -152,6 +156,56 @@ struct FillOpDeGeneralizationPattern
   }
 };
 
+// From linalg.generic to TPP brgemm (VNNI).
+struct TppBrgemmDeGeneralizationPattern
+    : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  bool isTppVnniOp(linalg::GenericOp linalgOp) const {
```
Comment (rengolin) [inline lib/TPP/LinalgDeGeneralize.cpp:164]: This looks like it could be in a more generic place?

File: lib/TPP/LinalgDeGeneralize.cpp [hunk 2]
```diff
@@ -152,6 +156,56 @@ struct FillOpDeGeneralizationPattern
...
+            .output(MatchOne(0), HasMap(EqualsTo(mapList[2])))
+            .region(MatchOne(0), WithOpChain<arith::MulFOp, arith::AddFOp>(
+                                     /*captures=*/nullptr));
+    return matmulMatcher.match(linalgOp);
+  }
+
+  LogicalResult matchAndRewrite(linalg::GenericOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    if (!isTppVnniOp(linalgOp))
+      return failure();
```
Comment (rengolin) [inline lib/TPP/LinalgDeGeneralize.cpp:199]: Can't the non-VNNI variant be like this as well?

**Review by rengolin — COMMENTED** [review]
Nice!

File: lib/TPP/LinalgDeGeneralize.cpp (hunk 1 above)
Comment (chelini) [reply lib/TPP/LinalgDeGeneralize.cpp]: I moved in TppUtils.h. However I think we should pull out all the `isXXXOps` from TppUtils as we are going to use them also when converting linalg to xsmm, and including `TppUtils.h` looks borken.

File: lib/TPP/LinalgDeGeneralize.cpp (hunk 2 above)
Comment (chelini) [reply lib/TPP/LinalgDeGeneralize.cpp:166]: the non-VNNI variant is `BatchReduceOpDeGeneralizationPattern`. I had to keep `mlir::linalgx::rewriteToBRGemmOp` as it is used in transform. I have a patch to clean this up, if you want I can queue it here.

**Pull request merged**